Asset Tracker v2: Rename GPS to GNSS and add GNSS module documentation (#6392)
nrfconnect/sdk-nrf: applications/asset_tracker_v2/doc/gnss_module.rst

.. _asset_tracker_v2_gnss_module:

GNSS module
###########

.. contents::
   :local:
   :depth: 2

The GNSS module controls the GNSS functionality of the nRF9160.

> It can be used to acquire the location of the device, either in the form of events containing a position, velocity and time (PVT) structure or NMEA sentences.

Reviewer: Maybe provide links to online docs or nordic docs on GGA, NMEA, and PVT? So that users can do a quick reference/doc lookup if they are curious.

Features
********

This section documents the various features implemented by the module.

GNSS control
============

The module uses the :ref:`nrfxlib:gnss_interface` to communicate with the nRF9160 modem and control its GNSS functionality.

> A GNSS search is started when the module receives an ``APP_EVT_DATA_GET`` event and ``GNSS`` type is listed in the event's list over data that shall be sampled.

Reviewer: Could refer to the variable name here so that it's clear where to look.

The default GNSS configuration is as follows:
* Single-fix mode, where the GNSS will search for a defined interval until it acquires a position fix or it times out.

> * Timeout set to either the value configured at build time, or to the dynamically updated value set by the cloud side.

Reviewer: Could also mention how the gnss module receives its configuration.

Module internals
================

The GNSS module is a thread-less module in terms of message processing.
However, the GNSS module differs from the other modules at this point.
It has a thread dedicated to processing GNSS events from the modem library, as opposed to processing messages like other modules with threads do.
The thread is used to offload data processing from the GNSS callbacks coming from the modem library, as these are called from interrupt context and no time consuming tasks should be performed there.

All incoming events from other modules are handled in the context of the event manager callback, as they all complete fast enough to not require a separate thread.

.. _configuration_options:

Configuration options
*********************

To enable the GNSS module, you must set the following option:

 * :kconfig:`CONFIG_GNSS_MODULE`

The module can report data either as a structure of position, velocity and time (PVT) information, or as an NMEA string of GGA type.
To select which format to use, configure on of the two options below:

>  * :kconfig:`CONFIG_GNSS_MODULE_PVT` - Include position, velocity and time (PVT) information in GNSS_EVT_DATA_READY events that are emitted when the GNSS acquires a position fix.
> 
>  * :kconfig:`CONFIG_GNSS_MODULE_NMEA` - Include NMEA string of GGA type in GNSS_EVT_DATA_READY events that are emitted when the GNSS acquires a position fix.

Reviewer 2: ```suggestion
 * :kconfig:`CONFIG_GNSS_MODULE_PVT` - Include position, velocity and time (PVT) information in the ``GNSS_EVT_DATA_READY`` events that are sent when the GNSS acquires a position fix.
 * :kconfig:`CONFIG_GNSS_MODULE_NMEA` - Include an NMEA string of the GGA type in the ``GNSS_EVT_DATA_READY`` events that are sent when the GNSS acquires a position fix.
```

> The GNSS module can set up the modem and LNA to use either onboard or external GNSS antenna using on of the two options below:

Reviewer 2: ```suggestion
The GNSS module can set up the modem and LNA to use either an onboard or external GNSS antenna using one of the following two options:
```

>  * :kconfig:`CONFIG_GNSS_MODULE_ANTENNA_ONBOARD` - Send antenna configurations to the modem when you compile for Thingy:91 or nRF9160 DK.
> 
>  * :kconfig:`CONFIG_GNSS_MODULE_ANTENNA_EXTERNAL` - This will configure the target for using an external GNSS antenna.

Reviewer 2: ```suggestion
 * :kconfig:`CONFIG_GNSS_MODULE_ANTENNA_ONBOARD` - Send antenna configurations to the modem when you compile for Thingy:91 or nRF9160 DK.
 * :kconfig:`CONFIG_GNSS_MODULE_ANTENNA_EXTERNAL` - This configures the target to use an external GNSS antenna.
```

If P-GPS is used, the last known position from a GNSS fix can be stored and injected to the modem together with the relevant ephemeris.

> This may, depending on the use-case and the device's movements, reduce the time to fix.
> To enable this feature, use the option below:

Reviewer 2: ```suggestion
This may, depending on the use case and the device's movements, reduce the time to fix.
To enable this feature, use the following option:
```

  * :kconfig:`CONFIG_GNSS_MODULE_PGPS_STORE_LOCATION`

For more information on P-GPS, see :ref:`lib_nrf_cloud_pgps`.

Module states
*************

The GNSS module has an internal state machine with the following states:

  * ``STATE_INIT`` - The initial state of the module in which it awaits the modem to be initialized and receive the GNSS configuration.
  * ``STATE_RUNNING`` - The module has performed all required initialization and can respond to requests to start GNSS searches. The running state has two sub-states:
    * ``SUB_STATE_SEARCH`` - A GNSS saerch is ongoing.

>     * ``SUB_STATE_IDLE`` - The module is idling, and able to respond to a request to start GNSS search.
>   * ``STATE_SHUTDOWN`` - The module has been shut down after receiving instruction to do so from the util module.

Reviewer 2: ```suggestion
    * ``SUB_STATE_IDLE`` - The module is idling and can respond to a request to start a GNSS search.
  * ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request from the util module.
```

> State transitions happen based on input from other modules, such as the app module, data module, util module and modem module.

Reviewer 2: ```suggestion
State transitions take place based on input from other modules, such as the app module, data module, util module and modem module.
```

Module events
*************

> The :file:`asset_tracker_v2/src/events/gnss_module_event.h` header file contains a list of the events emitted by the GNSS module and associated payload.

Reviewer 2: ```suggestion
The :file:`asset_tracker_v2/src/events/gnss_module_event.h` header file contains a list of the events sent by the GNSS module with the associated payload.
```
